### ISSUE TITLE
NAs produced in R-UDF

### ISSUE THREAD

**Issue by przell** [Open-EO/openeo-r-udf#12]
Hi Florian, 
we have encountered a situation where NAs are created in the R code. Then in the resulting hypercube.json they are written as character "NA". This cannot be handled by rasdaman. And is (as you have mentioned last time) also not correct to mix characters and numericals in one array.
Is there an R intrinsic numeric that is assigned as NA when writing to disc? 
Has the hypercube or new exchange format thought about this NA/NaN/NULL topic?
How is the python client handling this?
Somehow the information which value is NA should be provided via the data exchange format.
We can also raise the topic in the next dev meeting.
Best, 
Peter

**Comment by flahn** [comment]
Hi Peter,

sorry for the late reply, I must have overlooked the notification. There is an option to translate NA -> null during the conversion into JSON. As your input JSON `null` will be treated as NA then. Is this the intended behavior?

**Comment by m-mohr** [comment]
Just copying this from a mail conversation:

The API itself uses null for NA and doesn't allow transmitting NaN, although processes can handle NaN/infinite etc using the is_* processes. So it's a bit diferent for the API as it doesn't need to transport actual raw values and can assign null multiple meanings depending on the context. The issue here is that JSON can't fully represent IEEE754 floating point numbers, but that seems a very cruecial thing for transporting raw data, which is usually numeric in our case. So you are always loosing some information in JSON as you can only translate NaN/infinite/null to a single value. Therefore, it doesn't seem to be a good idea overall to transport the data in JSON format. There's probably better formats to work with, but if JSON should still be a thing, I'd propose to use objects as a replacement as usually our raw data is scalars and so there are no conflicts with objects. Could be simply something like { NaN: 1}, { inf: 1 }, { inf: -1 }, { NA: 1 } or whatever is relatively small (for transport size).

Edit: Messagepack supports NaN/infinity and there you could probably use null as NA.

**Comment by flahn** [comment]
Or to emphasize this. JSON only supports `null`. Which means NA, NaN and NULL in R is mapped to null in JSON. For R-UDF version 0.2.x (API version 0.1pre-alpha) we will have no alternatives in this.

However, in version > 0.3 ( API version 0.1 "the new one") we might define some special parameters that we exchange in the "server context" part - like "na_value" or "nan_value" or something alike.

For example:
```json
{
  ... ,
  "data" : {
    "proj": {},
    "user_context": {},
    "server_context": {
      "na_value": -3.4e+038,
      "nan_value": 3.4e+038
    },
    ...
  }
}
```

The values and the parameters, of course, have to be discussed and agreed upon.

**Comment by GreatEmerald** [comment]
This makes sense, and would work well for floats, but I'm not sure about integers; for a byte, you might need to allow the users to have some control over the NA value, because both 0 and 255 are pretty useful values that can mean a lot of things. It would be bad if we'd run a process over UDF and find out that suddenly, all the zeroes in the input come out as NA in the output...

**Comment by m-mohr** [comment]
I don't think it's a good idea to use numerical values as placeholder. That is just a workaround with less conflict potential, but still "limits" the range usable for "normal" data. I think my "object" approach above is more robust, but overall I'd try to ditch JSON in the long run and use something else (e.g. netCDF, messagepack or whatever more suited for raw/binary data exchange).

@GreatEmerald: JSON doesn't know integers at all, everything is float ("number"). So 3.4e+038 can still be used for 0-255 integers. The server would convert it properly.

**Comment by edzer** [comment]
@GreatEmerald if you have UInt1 data _with_ NA values there is no other option than to sacrifice one value for NA. I don't think that openEO goes as far as worrying with the representation on disc, but assumes there is some number for each pixel, and something that flags NAs.

**Comment by edzer** [comment]
(... or, what GEE does, keep a binary mask layer with everything around)

**Comment by flahn** [comment]
Just what I was thinking, Edzer. Apart from the context use. You can always create a NA mask as an additional attribute of a `stars` object. The output can have multiple attributes (e.g. bands), at least the UDF API allows us to do this. Whether the back-end can handle this, I don't know.

**Comment by GreatEmerald** [comment]
@m-mohr I thought that may be the case, thanks! I'm not entirely certain what you mean by an object approach, though; I thought you meant the same as what @flahn suggested?

@edzer Yes, indeed, hence why I said that it might be necessary for users to be able to control which value is sacrificed, if any. Just making sure that the users don't lose data without them knowing or being able to do anything about it...

**Comment by m-mohr** [comment]
@GreatEmerald I'm just talking about the transport layer between UDF server and back-end. For JSON, you could use something like {NA: true} or so to mimic NA values. For all other cases it's likely not a problem. It's similar to @flahn's approach, but may not conflict with any other data and covers all types of "special" numbers (NA, NaN, +/-infinite). Best would be to just abandon JSON though...

**Comment by GreatEmerald** [comment]
@m-mohr So in this example, does it mean that all entries that are `true` in the JSON would be considered as `NA` by the backend, or does it mean that all the values that are to be considered `NA` would need to be replaced by an object that is `{NA: true}` in the JSON itself?

**Comment by m-mohr** [comment]
> @m-mohr does it mean that all the values that are to be considered `NA` would need to be replaced by an object that is `{NA: true}` in the JSON itself?

This

**Comment by flahn** [comment]
@przell To get back to the start of this issue, it should be fixed in version v0.2.1 ( c702eaa) and similarly in the v0.3.1 for the new UDF API. Please checkout and build the dockers image with your additions for libraries and so on. Maybe report back to me if the fix works for you.

For a quick solution I'm mapping everything to null in the JSON. We should keep up with this issue of NA, NaN and other special numerical events, but maybe in a separate discussion in the UDF repo, because it should also concern the python domain. The object approach suggested by @m-mohr  seems to me the most flexible, non-data blocking solution. However this limits us to the JSON / message pack exchange. Because file formats use the flagging / data blocking approach as far as I know, but correct me if I'm wrong.

**Comment by m-mohr** [comment]
Good point about other file formats... Is there any file format that actually handles this better?

**Comment by GreatEmerald** [comment]
I see, though it becomes even less efficient to store "NA" as "{NA: true}"; keep in mind that e.g. BFAST Monitor outputs NA *by default*, outputting something other than NA only in the cases where a break in the time series has been detected. So in homogeneous areas you can expect the rasters to be almost nothing but NA values.

**Comment by m-mohr** [comment]
That is true, but we can't avoid this in JSON. It would be great to have an alternative file format for transport (we don't store here), but that's still to be found/defined. We can probably shorten `{"NA":true}` to `{"N":1}` or so (7 chars) and then it is actually shorter in size than `3.4e+038` (8 chars) in the other example. Otherwise you have to go back to some replacement value such as `0`, which limits the usability of your data range and users have to specify actual values for NA etc, which is bothersome. Best would still be to abandon JSON and come up with a better suited file format. Transport between UDF server and back-end should be lossless and not be exposed to the user through such additional "placeholder" settings.
